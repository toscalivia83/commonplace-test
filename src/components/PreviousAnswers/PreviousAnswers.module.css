.answerContainer {
  border: 1px solid black;
  margin: 10px 250px;
}

.answerItem {
  border: 0.5px solid grey;
  margin: 5px;
}