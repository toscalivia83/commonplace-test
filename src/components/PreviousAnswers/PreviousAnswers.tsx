import React from "react";
import { AnswerResponse } from "../../types/app";

import styles from "./PreviousAnswers.module.css";

const PreviousAnswers = ({ answers }: Props): React.ReactElement => {
  return <div>
    {answers.length && answers.map((answer, index) => 
      <div key={index} className={styles.answerContainer}>
        <p>Email Address: {answer.emailAddress}</p>
        <p>Option chosen:{answer.optionChosen}</p>
        <p>Comments: {answer.comments}</p>
      </div>
    )}
    {!answers.length && "No answers yet :)"}
  </div>;
};

export default PreviousAnswers;

interface Props {
  answers: AnswerResponse[];
}