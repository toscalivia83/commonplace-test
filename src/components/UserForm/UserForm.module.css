.field {
 display: block; 
}

.userForm {
  max-width: 700px;
  justify-content: center;
  display: flex;
  flex-direction: column;
  margin: 0 auto;
}

.formItem {
  display: flex;
  flex-direction: column;
  margin: 0 0 10px 0;
}

.label {
  text-align: left;
  margin: 5px;
}

.inputField {
  border: 1px solid grey;
  background-color: white;
  border-radius: 4px;
  padding: 8px 10px 6px 12px;
  font-size: 16px;
  line-height: 30px;
  vertical-align: middle;
  -webkit-appearance: none;
  min-width: 0;
}

.textarea {
  background: #fff;
  border: 1px solid grey;
  border-radius: 5px;
  font-size: 16px;
  line-height: 20px;
  padding: 12px 16px;
  resize: none;
}

.select {
  background: none;
  border: 1px solid grey;
  font-size: 16px;
}