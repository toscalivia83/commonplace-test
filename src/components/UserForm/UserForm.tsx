import React, { useState } from "react";
import EmailValidator from "email-validator";
import { UserFormDetails } from "../../types/app";
import styles from "./UserForm.module.css";

const validateText = (value: string): boolean => {
  const trimmedValue = value.trim();
  return Boolean(trimmedValue);
};

const isValidEmail = (email: string): boolean => {
  return validateText(email)
    && EmailValidator.validate(email);
};

const getErrorMessages = (userFormDetails: UserFormDetails): string[] => {
  const errorMessages = [];
  if (!validateText(userFormDetails.comments)) {
    errorMessages.push("Please enter valid comments");
  }
  if (!isValidEmail(userFormDetails.emailAddress)) {
    errorMessages.push("Please enter valid email address");
  }
  if (!validateText(userFormDetails.optionChosen)) {
    errorMessages.push("Select an option");
  }
  return errorMessages;
};

const UserForm = ({ options, handleSubmit }: UserFormProps): React.ReactElement => {
  const [userFormDetails, setUserFormDetails] = useState({ emailAddress: "", optionChosen: "", comments: "" });
  const [errorMessages, setErrorMessages] = useState([] as string[]);

  const handleSubmitClick = async (event: React.MouseEvent<HTMLButtonElement>): Promise<void> => {
    event.preventDefault();
    const errorMessages = getErrorMessages(userFormDetails);
    setErrorMessages(errorMessages);
    if (errorMessages.length === 0) {
      await handleSubmit(userFormDetails);
    }
  };

  const onChange = (fieldName: string, fieldValue: string): void => {
    setUserFormDetails({ ...userFormDetails, [fieldName]: fieldValue });
  };

  const onOptionSelect = (optionChosen: string): void => {
    setUserFormDetails({ ...userFormDetails, optionChosen });
  };

  return (
    <div>
      <form className={styles.userForm}>
        <div className={styles.formItem}>
          <label className={styles.label}>Email address</label>
          <input
            name="emailAddress"
            type="email"
            placeholder="Please type a valid email address..."
            value={userFormDetails.emailAddress}
            onChange={(event: React.FormEvent<HTMLInputElement>): void => onChange("emailAddress", event.currentTarget.value)}
            className={styles.inputField}/>
        </div>

        <div className={styles.formItem}>  
          <label className={styles.label}>What could you do if you weren t self isolating?</label>
          <select onChange={(evt): void => onOptionSelect(evt.target.value)} className={styles.select}>
            <option
              value={""}
              style={{ display: "none" }}
            ></option>
            {options.map((option, index) =>
              <option key={index} value={option}>{option}</option>
            )}
          </select>
        </div>
        
        <div className={styles.formItem}>
          <label className={styles.label}>Do you have any comments on the above?</label>
          <textarea
            onChange={(event): void => onChange("comments", event.currentTarget.value)}
            className={styles.textarea}></textarea>
        </div>
      </form>
      <button onClick={handleSubmitClick}>Submit your answers</button>
      {errorMessages.length > 0 && errorMessages.map((errorMessage, index) => <div key={index}>{errorMessage}</div>)}
    </div>
  );
};

export default UserForm;

interface UserFormProps {
  options: string[];
  handleSubmit: (userFormDetails: UserFormDetails) => Promise<void>;
}