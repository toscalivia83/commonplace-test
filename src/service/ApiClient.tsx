import { UserFormDetails, AnswerResponse, OptionsResponse } from "../types/app";
import { HttpError, BaseResponse } from "../types/http";

export class ApiClient {
  public readonly host: string;

  public constructor(baseUrl: string) {
    this.host = baseUrl;
  }

  public async getUserAnswers(): Promise<BaseResponse<AnswerResponse[]>> {
    const uri = `${this.host}/api/answers`;
    const response = await fetch(uri);

    try  {
      const responseJson = await response.json();
      return {
        statusCode: response.status,
        result: responseJson
      };
    } catch (e) {
      const error: HttpError = new Error(`request for ${uri} failed`);
      error.notFound = response.status === 404;
      error.status = response.status;
      throw error;
    }
  }

  public async getOptions(): Promise<BaseResponse<OptionsResponse>> {
    const uri = `${this.host}/api/options`;
    const response = await fetch(uri);

    try  {
      const responseJson = await response.json();
      return {
        statusCode: response.status,
        result: responseJson
      };
    } catch (e) {
      const error: HttpError = new Error(`request for ${uri} failed`);
      error.notFound = response.status === 404;
      error.status = response.status;
      throw error;
    }
  }

  public async postUserForm(userForm: UserFormDetails): Promise<BaseResponse<string>> {
    const requestOptions = {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Accept": "application/json"
      },
      body: JSON.stringify(userForm)
    };
    const uri = `${this.host}/api/answer`;
    const response = await fetch(uri, requestOptions);

    try  {
      const responseText = await response.text();
      return {
        statusCode: response.status,
        result: responseText
      };
    } catch (e) {
      const error: HttpError = new Error(`request for ${uri} failed`);
      error.notFound = response.status === 404;
      error.status = response.status;
      throw error;
    }
  }
}

export default new ApiClient(`${window.location.protocol}//${window.location.host}`);