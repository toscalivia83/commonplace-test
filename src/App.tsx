import React, { useEffect, useState } from "react";
import "./App.css";
import UserForm from "./components/UserForm/UserForm";
import ApiClient from "./service/ApiClient";
import { UserFormDetails, CurrentPage, AnswerResponse } from "./types/app";
import PreviousAnswers from "./components/PreviousAnswers/PreviousAnswers";

const App = (): React.ReactElement => {
  const [options, setOptions] = useState([] as string[]);
  const [previousUserAnswers, setPreviousUserAnswers] = useState([] as AnswerResponse[]);
  const [currentPage, setCurrentPage] = useState(CurrentPage.UserForm);

  const handleSubmit = async (userForm: UserFormDetails): Promise<void> => {
    await ApiClient.postUserForm(userForm);
    setCurrentPage(CurrentPage.PreviousAnswers);
  };

  const renderPage = (currentPage: string): React.ReactElement => {
    switch(currentPage) {
      case CurrentPage.UserForm:
        return <UserForm options={options} handleSubmit={handleSubmit} />;
      case CurrentPage.PreviousAnswers:
        return <PreviousAnswers answers={previousUserAnswers} />;
      default:
        return <UserForm options={options} handleSubmit={handleSubmit} />;
    }
  };

  useEffect(() => {
    async function fetchData(): Promise<void> {
      const optionsResponse = await ApiClient.getOptions();
      setOptions(optionsResponse.result?.options || [] as string[]);
      const answersResponse = await ApiClient.getUserAnswers();
      setPreviousUserAnswers(answersResponse.result || [] as AnswerResponse[]);
    }

    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    fetchData();
  }, []);

  return (
    <div className="App">
      <h1>Common place test - survey</h1>
      {renderPage(currentPage)}
    </div>
  );
};

export default App;
