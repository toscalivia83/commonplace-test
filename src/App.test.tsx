/* eslint-disable @typescript-eslint/unbound-method */
import React from "react";
import { mount, ReactWrapper } from "enzyme";
import ApiClient from "./service/ApiClient";
import App from "./App";
import UserForm from "./components/UserForm/UserForm";
import { act } from "@testing-library/react";
import PreviousAnswers from "./components/PreviousAnswers/PreviousAnswers";

const CALLED_ONCE = 1;

const flushPromise = async (): Promise<unknown> => {
  return new Promise<unknown>((resolve: (value?: unknown) => void): NodeJS.Immediate => setImmediate(resolve));
};

describe("<App /> suite", () => {
  let wrapper: ReactWrapper;

  it("renders App correctly",  () => {
    wrapper = mount(<App />);
    expect(wrapper.find(UserForm).exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders App correctly with options displayed", async () => {
    ApiClient.getOptions = jest.fn().mockResolvedValue(Promise.resolve({
      statusCode: 200,
      result: { options: ["a", "b", "c", "d"] }
    }));
    ApiClient.getUserAnswers = jest.fn().mockResolvedValue(Promise.resolve({
      statusCode: 200,
      result: [{ "emailAddress":"asdfg2@wsedf.fr","optionChosen":"I would do the same","comments":"aswdfghfds" }]
    }));

    await act(async (): Promise<void> => {
      wrapper = mount(<App />);
    });

    await flushPromise();
    wrapper.update();

    expect(ApiClient.getOptions).toHaveBeenCalledTimes(CALLED_ONCE);
    expect(ApiClient.getUserAnswers).toHaveBeenCalledTimes(CALLED_ONCE);
    expect(wrapper.find(UserForm).exists()).toBeTruthy();
  });

  it("should post the valid user form and go to previous answers page", async () => {
    ApiClient.getOptions = jest.fn().mockResolvedValue(Promise.resolve({
      statusCode: 200,
      result: { options: ["a", "b", "c", "d"] }
    }));
    ApiClient.getUserAnswers = jest.fn().mockResolvedValue(Promise.resolve({
      statusCode: 200,
      result: [{ "emailAddress":"asdfg2@wsedf.fr","optionChosen":"I would do the same","comments":"aswdfghfds" }]
    }));
    ApiClient.postUserForm = jest.fn().mockResolvedValue(Promise.resolve({
      statusCode: 200,
      result: "sent"
    }));

    await act(async (): Promise<void> => {
      wrapper = mount(<App />);
    });

    wrapper.find("input[name='emailAddress']").simulate("change", { target: { value: "test@domain.com" } });
    const select = wrapper.find("select");
    select.simulate("change", { target: { value: "a" } });
    wrapper.find("textarea").simulate("change", { target: { value: "test comment" } });

    wrapper.find("button").simulate("click");

    await act(async (): Promise<void> => {
      await flushPromise();
      wrapper.update();
    });

    expect(ApiClient.postUserForm).toHaveBeenCalledTimes(CALLED_ONCE);
    expect(wrapper.find(PreviousAnswers).exists()).toBeTruthy();
  });
});
