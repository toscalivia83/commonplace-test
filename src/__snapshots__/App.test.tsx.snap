// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> suite renders App correctly 1`] = `ReactWrapper {}`;
